src/cargo/core/compiler/build_context/target_info.rs

```diff
@@ -748,3 +751,69 @@ impl RustcTargetData {
         self.target_config(kind).links_overrides.get(lib_name)
     }
 }
+
+/// Structure used to deal with Rustdoc fingerprinting
+#[derive(Debug, Serialize, Deserialize)]
+pub struct RustDocFingerprint {
+    rustc_vv: String,
+}
+
+impl RustDocFingerprint {
+    /// Read the `RustDocFingerprint` info from the fingerprint file.
+    fn read<'a, 'cfg>(cx: &Context<'a, 'cfg>) -> CargoResult<Self> {
+        let rustdoc_data = paths::read(
+            &cx.files()
+                .host_root()
+                .join(".rustdoc_fingerprint")
+                .with_extension("json"),
+        )?;
+        serde_json::from_str(&rustdoc_data).map_err(|e| anyhow::anyhow!("{:?}", e))
+    }
+
+    /// Write the `RustDocFingerprint` info into the fingerprint file.
+    fn write<'a, 'cfg>(&self, cx: &Context<'a, 'cfg>) -> CargoResult<()> {
+        paths::write(
+            &cx.files()
+                .host_root()
+                .join(".rustdoc_fingerprint.json")
+                .with_extension("json"),
+            serde_json::to_string(&self)?.as_bytes(),
+        )
+    }
+
+    /// This function checks whether the latest version of `Rustc` used to compile this
+    /// `Workspace`'s docs was the same as the one is currently being used in this `cargo doc`
+    /// call.
+    ///
+    /// In case it's not, it takes care of removig the `doc/` folder as well as overwriting
+    /// the rustdoc fingerprint info in order to guarantee that we won't end up with mixed
+    /// versions of the `js/html/css` files that `Rustc` autogenerates which do not have
+    /// any versioning.
+    pub fn check_rustdoc_fingerprint<'a, 'cfg>(cx: &Context<'a, 'cfg>) -> CargoResult<()> {
+        let actual_rustdoc_target_data = RustDocFingerprint {
+            rustc_vv: cx.bcx.rustc().verbose_version.clone(),
+        };
+        let doc_dir = cx.files().host_root().join("doc");
```

Comment:
Docs are not always built in the host directory.  If you run `cargo doc --target x86_64-unknown-linux-gnu`, then the docs will be in the target-specific directory.

Reply:
Hmm that was a good catch! I completely forgot. I added a mechanism to collect the targets in the config and added the corresponding paths to the `check_rustdoc_fingerprint` logic.

Thanks!